components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/health/GatewayHealthMonitor.java

```diff
@@ -0,0 +1,294 @@
+                    scheduler.shutdownNow();
+                }
+            } catch (InterruptedException e) {
+                scheduler.shutdownNow();
+                Thread.currentThread().interrupt();
+            }
+        }
+    }
+
+    private String buildRegistrationPayload(String gwid) {
+        long millis = java.time.Instant.now().toEpochMilli();
+        // if gwid is null or empty we generate a new one using random UUID
+        if (gwid == null || gwid.isEmpty()) {
+            gwid = java.util.UUID.randomUUID().toString();
+        }
+        String ipAddress = APIUtil.getHostAddress();
+
+        ObjectMapper mapper = new ObjectMapper();
+        ObjectNode node = mapper.createObjectNode();
+        node.put(APIConstants.GatewayNotification.PAYLOAD_TYPE, APIConstants.GatewayNotification.PAYLOAD_TYPE_REGISTER);
+        node.put(APIConstants.GatewayNotification.GATEWAY_ID, gwid);
+        node.put(APIConstants.GatewayNotification.TIMESTAMP, millis);
+        
+        ObjectNode gatewayProperties = mapper.createObjectNode();
+        gatewayProperties.put("ipAddress", ipAddress);
+        node.set(APIConstants.GatewayNotification.GATEWAY_PROPERTIES, gatewayProperties);
+        
+        ArrayNode environmentLabels = mapper.createArrayNode();
+        for (String label : gatewayArtifactSynchronizerProperties.getGatewayLabels()) {
+            environmentLabels.add(label);
+        }
+        node.set(APIConstants.GatewayNotification.ENVIRONMENT_LABELS, environmentLabels);
+        
+        return node.toString();
+    }
+
+    private String buildHeartbeatPayload(String gwid) {
+        long millis = java.time.Instant.now().toEpochMilli();
+        
+        ObjectMapper mapper = new ObjectMapper();
+        ObjectNode node = mapper.createObjectNode();
+        node.put(APIConstants.GatewayNotification.PAYLOAD_TYPE, APIConstants.GatewayNotification.PAYLOAD_TYPE_HEARTBEAT);
+        node.put(APIConstants.GatewayNotification.GATEWAY_ID, gwid);
+        node.put(APIConstants.GatewayNotification.TIMESTAMP, millis);
+        
+        return node.toString();
+    }
+
+    private String notifyGateway(String payload) {
+        String endpoint = getServiceURL() + "/notify-gateway";
+        try {
+            HttpResponse response = executePost(endpoint, payload);
+            int statusCode = response.getStatusLine().getStatusCode();
+            String responseBody = new String(response.getEntity().getContent().readAllBytes(), StandardCharsets.UTF_8);
+            log.debug("/notify-gateway called. Status: " + statusCode + ", Response: " + responseBody);
+            return responseBody;
+        } catch (IOException e) {
+            log.error("Error occurred while calling /notify-gateway", e);
+            return null;
+        }
+    }
+
+    private String extractGatewayIdFromResponse(String responseBody) {
+        int idx = responseBody.indexOf("\"" + APIConstants.GatewayNotification.GATEWAY_ID + "\"");
```

Comment:
Instead of processing the string, convert this  JSON to Java Object using Gson. See sample https://github.com/wso2/carbon-apimgt/blob/master/components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/EndpointCertificateDeployer.java#L98